src/model/warning.ts

```diff
@@ -12,12 +12,114 @@
  * http://polymer.github.io/PATENTS.txt
  */
 
+import * as chalk from 'chalk';
+
+import {ParsedDocument} from '../parser/document';
+import {underlineCode} from '../warning/code-printer';
+
 import {SourceRange} from './source-range';
-export interface Warning {
+
+export interface WarningInit {
   readonly message: string;
   readonly sourceRange: SourceRange;
   readonly severity: Severity;
   readonly code: string;
+  readonly parsedDocument: ParsedDocument<any, any>|null;
+}
+export class Warning {
+  readonly code: string;
+  readonly message: string;
+  readonly sourceRange: SourceRange;
+  readonly severity: Severity;
+
+  private readonly _parsedDocument: ParsedDocument<any, any>|null;
+
+  constructor(init: WarningInit) {
+    ({
+      message: this.message,
+      sourceRange: this.sourceRange,
+      severity: this.severity,
+      code: this.code,
+      parsedDocument: this._parsedDocument,
+    } = init);
+  }
+
+  toString(options: Partial<WarningStringifyOptions> = {}): string {
+    const opts:
+        WarningStringifyOptions = {...defaultPrinterOptions, ...options};
+    const colorize = opts.color ? this._severityToColorFunction(this.severity) :
+                                  (s: string) => s;
+    const severity = this._severityToString(colorize);
+
+    if (!this.sourceRange || !this._parsedDocument) {
```

Comment:
why can't you do this check on construction? It'll be closer to the real error and a stack might be useful.